easyunfold/procar.py

```diff
@@ -32,85 +38,198 @@ def __init__(self, fobj_or_path=None, is_soc=False):
 
         # Read the PROCAR
-        if isinstance(fobj_or_path, (str, Path)):
-            with open(fobj_or_path, encoding='utf-8') as fhandle:
-                self._read(fhandle)
-        else:
-            self._read(fobj_or_path)
+        if isinstance(fobjs_or_paths, (str, Path)):
+            fobjs_or_paths = [fobjs_or_paths]
+        self.read(fobjs_or_paths)
 
-    def _read(self, fobj):
+    def _read(self, fobj, parsed_kpoints=None):
         """Main function for reading in the data"""
+        if parsed_kpoints is None:
+            parsed_kpoints = set()
 
-        # First sweep - found the number of kpoints and the number of bands
+        # First sweep - find the number of kpoints and the number of bands
         fobj.seek(0)
-        self.header = fobj.readline()
+        _header = fobj.readline()
         # Read the NK, NB and NIONS that are integers
-        self.nkpts, self.nbands, self.nion = [int(token) for token in re.sub(r'[^0-9]', ' ', fobj.readline()).split()]
-        # Number of projects and their names
-        nproj = None
-        self.proj_names = None
-
-        for line in fobj:
-            if re.match(r'^ion.*tot', line):  # only the first "ion" line, in case of LORBIT >= 12
-                nproj = len(line.strip().split()) - 2
-                self.proj_names = line.strip().split()[1:-1]
-                break
+        _total_nkpts, nbands, nion = [int(token) for token in re.sub(r'[^0-9]', ' ', fobj.readline()).split()]
+        if nion != self.nion:
+            raise ValueError(f'Mismatch in number of ions in PROCARs supplied: ({nion} vs {self.nion})!')
 
         # Count the number of data lines, these lines do not have any alphabets
-        proj_data = []
-        energies = []
-        occs = []
-        kvecs = []
-        kweights = []
+        proj_data, energies, kvecs, kweights, occs = [], [], [], [], []
+        tot_count = 0  # count the instances of lines starting with "tot" -> (4 + 1) * nbands * nkpts for SOC calcs
         fobj.seek(0)
-        for line in fobj:
-            if not re.search(r'[a-zA-Z]', line) and line.strip() and len(line.strip().split()) - 2 == nproj:
-                # only parse data if nproj is expected length, in case of LORBIT >= 12
-                proj_data.append([float(token) for token in line.strip().split()[1:-1]])
-            elif line.startswith('band'):
-                tokens = line.strip().split()
-                energies.append(float(tokens[4]))
-                occs.append(float(tokens[-1]))
-            elif line.startswith(' k-point'):
-                line = re.sub(r'(\d)-', r'\1 -', line)
-                tokens = line.strip().split()
-                kvecs.append([float(val) for val in tokens[-6:-3]])
-                kweights.append(float(tokens[-1]))
-        self.occs = np.array(occs)
-        self.kvecs = np.array(kvecs)
-        self.kweights = np.array(kweights)
-        self.eigenvalues = np.array(energies)
+
+        with fobj:
```

Comment:
We should not preemptively close the file object here, e.g. the stream may be used by other  functions later. 
Closing the file object can be moved to the `read` method instead, if the file object is opened by that method.